File: Source/AlphaTab/Importer/PartConfiguration.cs
using AlphaTab.Collections;
using AlphaTab.IO;
using AlphaTab.Model;

namespace AlphaTab.Importer
{
    class PartConfiguration
    {
        public class Part
        {
            public bool IsMultiRest { get; set; }

            public FastList<TrackConfiguration> Tracks
            {
                get;
                set;
            }

            public Part()
            {
                Tracks = new FastList<TrackConfiguration>();
            }
        }

        public class TrackConfiguration
        {
            public bool IsVisible { get; set; }
            public bool ShowSlash { get; set; }
            public bool ShowStandardNotation { get; set; }
            public bool ShowTablature { get; set; }
        }

        public FastList<Part> Parts { get; set; }
        public int ZoomLevel { get; set; }
        public int Layout { get; set; }

        public PartConfiguration()
        {
            Parts = new FastList<Part>();
        }

        public void Apply(Score score)
        {
            var staffIndex = 0;
            var trackIndex = 0;

            // the PartConfiguration is really twisted compared to how the score structure looks like. 
            // the first part typically contains the settings for the first staff of all tracks. 
            // but then there is 1 part with 1 track for each other staff of the tracks. 
            // So the structure in the PartConfig appears to be:
            // Parts[0].Tracks = { Track1-Staff1, Track2-Staff1, Track3-Staff1, Track4-Staff1, .. }
            // Parts[1].Tracks = { Track1-Staff2 }
            // Parts[2].Tracks = { Track2-Staff2 }
            // Parts[3].Tracks = { Track3-Staff2 }
            // Parts[4].Tracks = { Track4-Staff2 }
            //
            // even if a track has only 1 staff, there are 2 staff configurations stored.
            // I hope Arobas never changes this in the format as the PartConfiguration is not versionized. 

            foreach (var part in Parts)
            {
                foreach (var trackConfig in part.Tracks)
                {
                    if (trackIndex < score.Tracks.Count)
                    {
                        var track = score.Tracks[trackIndex];
                        if (staffIndex < track.Staves.Count)
                        {
                            var staff = track.Staves[staffIndex];
                            staff.ShowTablature = trackConfig.ShowTablature;
                            staff.ShowStandardNotation = trackConfig.ShowStandardNotation;
                        }
                    }

                    trackIndex++;
                    if (trackIndex >= score.Tracks.Count)
                    {
                        staffIndex++;
                        trackIndex = 0;
                    }
                }
            }
        }
    }

    class PartConfigurationParser
    {
        public PartConfiguration Configuration { get; private set; }

        public void Parse(byte[] partConfigurationData)
        {
            Configuration = new PartConfiguration();

            ParsePartConfiguration(partConfigurationData);
        }

        private void ParsePartConfiguration(byte[] partConfigurationData)
        {
            var readable = ByteBuffer.FromBuffer(partConfigurationData);
            var entryCount = readable.ReadInt32BE();

            for (int i = 0; i < entryCount; i++)
            {
                var part = new PartConfiguration.Part();
                Configuration.Parts.Add(part);
                part.IsMultiRest = readable.GpReadBool();

                var groupCount = readable.ReadInt32BE();
                for (int j = 0; j < groupCount; j++)
                {
                    var flags = readable.ReadByte();
                    // enable at least standard notation
                    if (flags == 0)
                    {
                        flags = 1;
                    }

                    part.Tracks.Add(new PartConfiguration.TrackConfiguration
                    {
                        ShowStandardNotation = (flags & 0x01) != 0,
                        ShowTablature = (flags & 0x02) != 0,
                        ShowSlash = (flags & 0x04) != 0,
                    });
                }
            }
        }
    }
}

File: src/importer/PartConfiguration.ts
import { GpBinaryHelpers } from '@src/importer/Gp3To5Importer';
import { ByteBuffer } from '@src/io/ByteBuffer';
import { IOHelper } from '@src/io/IOHelper';
import type { Score } from '@src/model/Score';
import type { Track } from '@src/model/Track';

// PartConfiguration File Format Notes.
// Based off Guitar Pro 8
// The file contains a serialized "Score View Collection" filled like this:
// There is always 1 ScoreView holding a TrackViewGroup for each Track contained in the file. This is the multi-track layout.
// Additionally there is 1 ScoreView individually for each track with only 1 TrackViewGroup of this Group.

// The Guitar Pro UI seem to update both the multi-track and the single-track layouts when changing the displayed staves.
// But technically it would support showing alternating staves in multi-track.

// For the multi-rest flag the respective TrackViewGroups need to be respected.

// File:
//    int32 (big endian) | Number of Score Views
//    ScoreView[]        | The individual score views
//    int32 (big endian) | The index to the currently active view

// ScoreView:
//    1 byte (boolean)   | "Multi Rest" - Whether multiple bars with rests should be combined (Bar > Multi Rest option)
//    int32 (big endian) | Number of following tems
//    TrackViewGroup[]   | The individual track view groups

// TrackViewGroup:
//    1 byte             | Track View Group Type Bitflag
//                       | 0th bit: showStandardNotation
//                       | 1th bit: showTablature
//                       | 2nd bit: showSlash
//                       | 3rd bit: numberedNotation (GP8 feature - jiǎnpǔ  aka Chinese Number Notation)
//                       | if no bits set -> activate standard notation
//

class PartConfigurationScoreView {
    public isMultiRest: boolean = false;
    public trackViewGroups: PartConfigurationTrackViewGroup[] = [];
}

class PartConfigurationTrackViewGroup {
    public showNumbered: boolean = false;
    public showSlash: boolean = false;
    public showStandardNotation: boolean = false;
    public showTablature: boolean = false;
}

export class PartConfiguration {
    public scoreViews: PartConfigurationScoreView[] = [];

    public apply(score: Score): void {
        // for now we only look at the first score view which seem to hold
        // the config for all tracks.
        if (this.scoreViews.length > 0) {
            let trackIndex = 0;

            score.stylesheet.multiTrackMultiBarRest = this.scoreViews[0].isMultiRest;

            for (const trackConfig of this.scoreViews[0].trackViewGroups) {
                if (trackIndex < score.tracks.length) {
                    const track: Track = score.tracks[trackIndex];
                    for (const staff of track.staves) {
                        staff.showTablature = trackConfig.showTablature;
                        staff.showStandardNotation = trackConfig.showStandardNotation;
                        staff.showSlash = trackConfig.showSlash;
                        staff.showNumbered = trackConfig.showNumbered;
                    }
                }
                trackIndex++;
            }

            for (let scoreViewIndex = 1; scoreViewIndex < this.scoreViews.length; scoreViewIndex++) {
                if (this.scoreViews[scoreViewIndex].isMultiRest) {
                    // lazy init
                    if (!score.stylesheet.perTrackMultiBarRest) {
                        score.stylesheet.perTrackMultiBarRest = new Set<number>();
                    }

                    trackIndex = scoreViewIndex - 1;
                    score.stylesheet.perTrackMultiBarRest!.add(trackIndex);
                }
            }
        }
    }

    public constructor(partConfigurationData: Uint8Array) {
        const readable: ByteBuffer = ByteBuffer.fromBuffer(partConfigurationData);

        const scoreViewCount: number = IOHelper.readInt32BE(readable);

        for (let i: number = 0; i < scoreViewCount; i++) {
            const scoreView = new PartConfigurationScoreView();
            this.scoreViews.push(scoreView);

            scoreView.isMultiRest = GpBinaryHelpers.gpReadBool(readable);

            const trackViewGroupCount: number = IOHelper.readInt32BE(readable);
            for (let j: number = 0; j < trackViewGroupCount; j++) {
                let flags: number = readable.readByte();
                // enable at least standard notation
                if (flags === 0) {
                    flags = 1;
                }
                const trackConfiguration = new PartConfigurationTrackViewGroup();
                trackConfiguration.showStandardNotation = (flags & 0x01) !== 0;
                trackConfiguration.showTablature = (flags & 0x02) !== 0;
                trackConfiguration.showSlash = (flags & 0x04) !== 0;
                trackConfiguration.showNumbered = (flags & 0x08) !== 0;
                scoreView.trackViewGroups.push(trackConfiguration);
            }
        }
    }

    public static writeForScore(score: Score): Uint8Array {
        const writer = ByteBuffer.withCapacity(128);

        const scoreViews: PartConfigurationScoreView[] = [
            new PartConfigurationScoreView() // Multi Track Score View
        ];

        scoreViews[0].isMultiRest = score.stylesheet.multiTrackMultiBarRest;

        for (const track of score.tracks) {
            const trackConfiguration = new PartConfigurationTrackViewGroup();
            // NOTE: unclear how multi staff settings are meant in this format
            // in the Guitar Pro UI there is no individual staff config
            trackConfiguration.showStandardNotation = track.staves[0].showStandardNotation;
            trackConfiguration.showTablature = track.staves[0].showTablature;
            trackConfiguration.showSlash = track.staves[0].showSlash;
            trackConfiguration.showNumbered = track.staves[0].showNumbered;

            scoreViews[0].trackViewGroups.push(trackConfiguration);

            const singleTrackScoreView = new PartConfigurationScoreView();
            singleTrackScoreView.isMultiRest = score.stylesheet.perTrackMultiBarRest?.has(track.index) === true;
            singleTrackScoreView.trackViewGroups.push(trackConfiguration);
            scoreViews.push(singleTrackScoreView);
        }

        IOHelper.writeInt32BE(writer, scoreViews.length);
        for (const part of scoreViews) {
            writer.writeByte(part.isMultiRest ? 1 : 0);
            IOHelper.writeInt32BE(writer, part.trackViewGroups.length);
            for (const track of part.trackViewGroups) {
                let flags = 0;
                if (track.showStandardNotation) {
                    flags = flags | 0x01;
                }
                if (track.showTablature) {
                    flags = flags | 0x02;
                }
                if (track.showSlash) {
                    flags = flags | 0x04;
                }
                if (track.showNumbered) {
                    flags = flags | 0x08;
                }
                writer.writeByte(flags);
            }
        }

        IOHelper.writeInt32BE(writer, 1 /* First Single Track Layout */);

        return writer.toArray();
    }
}
